File: EMS/EMSMO/Session.m
//
//  Session.m
//

#import "Session.h"
#import "Slot.h"


@implementation Session

@dynamic attachmentCollection;
@dynamic audience;
@dynamic body;
@dynamic favourite;
@dynamic format;
@dynamic href;
@dynamic language;
@dynamic level;
@dynamic roomName;
@dynamic slotName;
@dynamic speakerCollection;
@dynamic state;
@dynamic summary;
@dynamic title;
@dynamic conference;
@dynamic keywords;
@dynamic room;
@dynamic slot;
@dynamic speakers;
@dynamic videoLink;


- (NSString *)sanitizedTitle {
    NSCharacterSet *notAllowedChars = [[NSCharacterSet characterSetWithCharactersInString:@"abcdefghijklmnopqrstuvwxyz"] invertedSet];
    return [[[self.title lowercaseString] componentsSeparatedByCharactersInSet:notAllowedChars] componentsJoinedByString:@""];
}

- (NSString *) sectionTitle {
    
    static NSDateFormatter *dateFormatter;
    static NSDateFormatter *timeFormatter;

    static NSDateFormatter *dateParser;
    static NSDateFormatter *timeParser;

    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        
        dateFormatter = [[NSDateFormatter alloc] init];
        dateFormatter.locale = [NSLocale autoupdatingCurrentLocale];
        dateFormatter.dateFormat = @"EEEE";
        
        timeFormatter = [[NSDateFormatter alloc] init];
        dateFormatter.locale = [NSLocale autoupdatingCurrentLocale];
        timeFormatter.dateStyle = NSDateFormatterNoStyle;
        timeFormatter.timeStyle = NSDateFormatterShortStyle;

        dateParser = [[NSDateFormatter alloc] init];
        [dateParser setDateFormat:@"yyyy-MM-dd"];

        timeParser = [[NSDateFormatter alloc] init];
        [timeParser setDateFormat:@"HH:mm"];
    });

    NSArray *parts = [self.slotName componentsSeparatedByString:@" "];

    NSString *date = [[dateFormatter stringFromDate:[dateParser dateFromString:parts[0]]] uppercaseString];
    NSString *startTime = [timeFormatter stringFromDate:[timeParser dateFromString:parts[1]]];
    NSString *endTime = [timeFormatter stringFromDate:[timeParser dateFromString:parts[3]]];

    if (date == nil || startTime == nil || endTime == nil) {
        return self.slotName;
    }

    return  [NSString stringWithFormat:@"%@ %@ - %@", date, startTime, endTime];
}

@end

File: EMS/CoreData Model/Model/Session.swift
//
//  Session.swift
//  EMS
//
//  Created by Chris Searle on 04/05/16.
//  Copyright © 2016 Chris Searle. All rights reserved.
//

import Foundation
import CoreData

enum DateFormatterError : ErrorType {
    case InvalidFormat, InvalidDate
}

@objc(Session)
class Session: NSManagedObject {
    struct Token {
        static var token : dispatch_once_t = 0
    }
    
    struct Formatters {
        static var dateFormatter : NSDateFormatter? = nil
        static var timeFormatter : NSDateFormatter? = nil
    }
    
    struct Parsers {
        static var dateParser : NSDateFormatter? = nil
        static var timeParser : NSDateFormatter? = nil
    }


    func sanitizedTitle() -> String {
        let allowedChars = Set("abcdefghijklmnopqrstuvwxyz".characters)

        return String(self.title?.lowercaseString.characters.filter {allowedChars.contains($0)})
    }
    
    func sectionTitle() -> String {
        dispatch_once(&Token.token) {
            Formatters.dateFormatter = NSDateFormatter()
            Formatters.dateFormatter!.locale = NSLocale.autoupdatingCurrentLocale()
            Formatters.dateFormatter!.dateFormat = "EEEE"
            
            Formatters.timeFormatter = NSDateFormatter()
            Formatters.timeFormatter!.locale = NSLocale.autoupdatingCurrentLocale()
            Formatters.timeFormatter!.dateStyle = .NoStyle
            Formatters.timeFormatter!.timeStyle = .ShortStyle
            
            Parsers.dateParser = NSDateFormatter()
            Parsers.dateParser!.dateFormat = "yyyy-MM-dd"
            
            Parsers.timeParser = NSDateFormatter()
            Parsers.timeParser!.dateFormat = "HH:mm"
        }
        
        do {
            return try formatSectionName(self.slotName)
        } catch {
            return self.slotName!
        }
    }
    
    private func formatSectionName(name: String?)  throws -> String {
        guard let parts = name?.characters.split(" ")
            else { throw DateFormatterError.InvalidFormat }
        
        guard parts.count >= 4
            else { throw DateFormatterError.InvalidFormat }
        
        guard let dateStr = Parsers.dateParser!.dateFromString(String(parts[0]))
            else { throw DateFormatterError.InvalidDate }
        
        guard let startTimeStr = Parsers.timeParser!.dateFromString(String(parts[1]))
            else { throw DateFormatterError.InvalidDate }
        
        guard let endTimeStr = Parsers.timeParser!.dateFromString(String(parts[3]))
            else { throw DateFormatterError.InvalidDate }
        
        let date = Formatters.dateFormatter!.stringFromDate(dateStr).uppercaseString
        let startTime = Formatters.timeFormatter!.stringFromDate(startTimeStr)
        let endTime = Formatters.timeFormatter!.stringFromDate(endTimeStr)
        
        return "\(date) \(startTime) - \(endTime)"
    }
}
